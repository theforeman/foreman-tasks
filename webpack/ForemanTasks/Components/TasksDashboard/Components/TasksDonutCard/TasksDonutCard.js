import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Card } from 'patternfly-react';

import {
  TASKS_DASHBOARD_AVAILABLE_QUERY_MODES,
  TASKS_DASHBOARD_AVAILABLE_TIMES,
} from '../../TasksDashboardConstants';
import { getTimeText } from '../../TasksDashboardHelper';
import { timePropType, queryPropType } from '../../TasksDashboardPropTypes';
import { TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS } from '../TasksDonutChart/TasksDonutChartConstants';
import { getFocusedOn } from '../TasksDonutChart/TasksDonutChartHelper';
import TasksDonutChart from '../TasksDonutChart/TasksDonutChart';
import './TasksDonutCard.scss';

const TasksDonutCard = ({
  title,
  data,
  query,
  wantedQueryState,
  className,
  time,
  updateQuery,
  ...props
}) => {
  const { LAST, OLDER } = TASKS_DASHBOARD_AVAILABLE_QUERY_MODES;
  const focusedOn = getFocusedOn(query, wantedQueryState, time);
  const onTotalClick = () => updateQuery({ state: wantedQueryState });

  return (
    <Card
      className={classNames('tasks-donut-card', className, {
        'selected-tasks-card':
          focusedOn !== TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.NORMAL &&
          focusedOn !== TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.NONE,
      })}
      {...props}
    >
      <Card.Title onClick={onTotalClick}>{title}</Card.Title>
      <Card.Body>
        <TasksDonutChart
          last={data.last}
          older={data.older}
          time={getTimeText(time)}
          focusedOn={focusedOn}
          onTotalClick={onTotalClick}
          onLastClick={() =>
            updateQuery({ state: wantedQueryState, mode: LAST, time })
          }
          onOlderClick={() =>
            updateQuery({ state: wantedQueryState, mode: OLDER, time })
          }
        />
      </Card.Body>
    </Card>
  );
};

TasksDonutCard.propTypes = {
  title: PropTypes.string,
  data: PropTypes.shape({
    last: PropTypes.number.isRequired,
    older: PropTypes.number.isRequired,
  }),
  time: timePropType,
  query: queryPropType,
  wantedQueryState: PropTypes.string,
  className: PropTypes.string,
  updateQuery: PropTypes.func,
};

TasksDonutCard.defaultProps = {
  title: '',
  data: { last: 0, older: 0 },
  query: {},
  wantedQueryState: '',
  time: TASKS_DASHBOARD_AVAILABLE_TIMES.H24,
  className: '',
  updateQuery: () => null,
};

export default TasksDonutCard;
