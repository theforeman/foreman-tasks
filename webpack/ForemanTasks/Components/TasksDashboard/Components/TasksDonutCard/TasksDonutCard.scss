.tasks-donut-card {
  .card-pf-title {
    text-align: center;
    font-size: 180%;
    cursor: pointer;
  }

  .card-pf-body {
    padding-left: 15px;
  }

  &.selected-tasks-card {
    .card-pf-title {
      font-weight: bold;
    }
  }
}
