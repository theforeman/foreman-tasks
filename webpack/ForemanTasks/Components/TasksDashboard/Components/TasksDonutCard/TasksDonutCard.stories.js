import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, number, text, select } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import { withCardsDecorator } from '../../../../../stories/decorators';

import {
  TASKS_DASHBOARD_AVAILABLE_QUERY_MODES,
  TASKS_DASHBOARD_AVAILABLE_TIMES,
} from '../../TasksDashboardConstants';
import TasksDonutCard from './TasksDonutCard';

storiesOf('TasksDashboard', module)
  .addDecorator(withKnobs)
  .addDecorator(withCardsDecorator)
  .add('TasksDonutCard', () => {
    const selectTime = select(
      'time',
      TASKS_DASHBOARD_AVAILABLE_TIMES,
      TasksDonutCard.defaultProps.time
    );

    const selectMode = select(
      'mode',
      { ...TASKS_DASHBOARD_AVAILABLE_QUERY_MODES, NONE: 'none' },
      'none'
    );

    return (
      <TasksDonutCard
        title={text('title', 'Some Title')}
        data={{
          last: number('data.last', 3),
          older: number('data.older', 5),
        }}
        time={selectTime}
        query={{
          state: text('query.state', 'some-state'),
          mode: selectMode,
          time: selectTime,
        }}
        wantedQueryState={text('wantedQueryState', 'wanted-state')}
        updateQuery={action('updateQuery')}
      />
    );
  });
