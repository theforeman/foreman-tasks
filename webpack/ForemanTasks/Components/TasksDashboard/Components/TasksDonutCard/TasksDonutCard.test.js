import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import {
  TASKS_DASHBOARD_AVAILABLE_TIMES,
  TASKS_DASHBOARD_AVAILABLE_QUERY_MODES,
} from '../../TasksDashboardConstants';
import TasksDonutCard from './TasksDonutCard';

const fixtures = {
  'render with minimal props': {},
  'render with props': {
    title: 'some title',
    time: TASKS_DASHBOARD_AVAILABLE_TIMES.WEEK,
    wantedQueryState: 'some-state',
    className: 'some-classname',
    data: { last: 3, older: 5 },
  },
  'render with total selected': {
    wantedQueryState: 'some-state',
    query: { state: 'some-state' },
  },
  'render with last selected': {
    wantedQueryState: 'some-state',
    time: TASKS_DASHBOARD_AVAILABLE_TIMES.WEEK,
    query: {
      state: 'some-state',
      mode: TASKS_DASHBOARD_AVAILABLE_QUERY_MODES.LAST,
      time: TASKS_DASHBOARD_AVAILABLE_TIMES.WEEK,
    },
  },
  'render with older selected': {
    wantedQueryState: 'some-state',
    time: TASKS_DASHBOARD_AVAILABLE_TIMES.WEEK,
    query: {
      state: 'some-state',
      mode: TASKS_DASHBOARD_AVAILABLE_QUERY_MODES.OLDER,
      time: TASKS_DASHBOARD_AVAILABLE_TIMES.WEEK,
    },
  },
};

describe('TasksDonutCard', () =>
  testComponentSnapshotsWithFixtures(TasksDonutCard, fixtures));
