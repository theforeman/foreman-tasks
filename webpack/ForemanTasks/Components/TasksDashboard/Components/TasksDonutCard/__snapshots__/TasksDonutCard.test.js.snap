// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksDonutCard render with last selected 1`] = `
<Card
  accented={false}
  aggregated={false}
  aggregatedMini={false}
  cardRef={null}
  className="tasks-donut-card selected-tasks-card"
  matchHeight={false}
>
  <CardTitle
    className=""
    onClick={[Function]}
  />
  <CardBody
    className=""
  >
    <TasksDonutChart
      className=""
      colorsPattern={
        Array [
          "#C315C7",
          "#0089C9",
        ]
      }
      focusedOn="last"
      last={0}
      older={0}
      onLastClick={[Function]}
      onOlderClick={[Function]}
      onTotalClick={[Function]}
      time="week"
    />
  </CardBody>
</Card>
`;

exports[`TasksDonutCard render with minimal props 1`] = `
<Card
  accented={false}
  aggregated={false}
  aggregatedMini={false}
  cardRef={null}
  className="tasks-donut-card"
  matchHeight={false}
>
  <CardTitle
    className=""
    onClick={[Function]}
  />
  <CardBody
    className=""
  >
    <TasksDonutChart
      className=""
      colorsPattern={
        Array [
          "#C315C7",
          "#0089C9",
        ]
      }
      focusedOn="normal"
      last={0}
      older={0}
      onLastClick={[Function]}
      onOlderClick={[Function]}
      onTotalClick={[Function]}
      time="24h"
    />
  </CardBody>
</Card>
`;

exports[`TasksDonutCard render with older selected 1`] = `
<Card
  accented={false}
  aggregated={false}
  aggregatedMini={false}
  cardRef={null}
  className="tasks-donut-card selected-tasks-card"
  matchHeight={false}
>
  <CardTitle
    className=""
    onClick={[Function]}
  />
  <CardBody
    className=""
  >
    <TasksDonutChart
      className=""
      colorsPattern={
        Array [
          "#C315C7",
          "#0089C9",
        ]
      }
      focusedOn="older"
      last={0}
      older={0}
      onLastClick={[Function]}
      onOlderClick={[Function]}
      onTotalClick={[Function]}
      time="week"
    />
  </CardBody>
</Card>
`;

exports[`TasksDonutCard render with props 1`] = `
<Card
  accented={false}
  aggregated={false}
  aggregatedMini={false}
  cardRef={null}
  className="tasks-donut-card some-classname"
  matchHeight={false}
>
  <CardTitle
    className=""
    onClick={[Function]}
  >
    some title
  </CardTitle>
  <CardBody
    className=""
  >
    <TasksDonutChart
      className=""
      colorsPattern={
        Array [
          "#C315C7",
          "#0089C9",
        ]
      }
      focusedOn="normal"
      last={3}
      older={5}
      onLastClick={[Function]}
      onOlderClick={[Function]}
      onTotalClick={[Function]}
      time="week"
    />
  </CardBody>
</Card>
`;

exports[`TasksDonutCard render with total selected 1`] = `
<Card
  accented={false}
  aggregated={false}
  aggregatedMini={false}
  cardRef={null}
  className="tasks-donut-card selected-tasks-card"
  matchHeight={false}
>
  <CardTitle
    className=""
    onClick={[Function]}
  />
  <CardBody
    className=""
  >
    <TasksDonutChart
      className=""
      colorsPattern={
        Array [
          "#C315C7",
          "#0089C9",
        ]
      }
      focusedOn="total"
      last={0}
      older={0}
      onLastClick={[Function]}
      onOlderClick={[Function]}
      onTotalClick={[Function]}
      time="24h"
    />
  </CardBody>
</Card>
`;
